torchao/dtypes/uintx/gemlite_layout.py

```diff
@@ -0,0 +1,372 @@
+
+    quant_kwargs = get_gemlite_quant_kwargs(bit_width, group_size)
+
+    layout = GemlitePackedLayout(
+        group_size=group_size,
+        bit_width=bit_width,
+        packing_bitwidth=packing_bitwidth,
+        contiguous=contiguous,
+    )
+    return to_affine_quantized_intx(
+        weight, **quant_kwargs, _layout=layout, use_hqq=use_hqq
+    )
+
+
+@dataclass(frozen=True)
+class GemlitePackedLayout(Layout):
+    group_size: Optional[int] = 64
+    bit_width: int = 4
+    packing_bitwidth: int = 8
+    contiguous: bool = None
+
+
+@register_layout(GemlitePackedLayout)
+class GemliteAQTTensorImpl(TensorCoreTiledAQTTensorImpl):
+    def __new__(
+        cls,
+        packed_weight: torch.Tensor,
+        scale: torch.Tensor,
+        zero_point: torch.Tensor,
+        gemlite_kwargs: Dict,
+        _layout: Layout,
+    ):
+        kwargs = {}
+        kwargs["device"] = packed_weight.device
+        kwargs["layout"] = (
+            kwargs.get("layout")
+            if kwargs.get("layout", False)
+            else packed_weight.layout
+        )
+        kwargs["dtype"] = packed_weight.dtype
+        kwargs["requires_grad"] = False
+        shape = packed_weight.shape
+        return torch.Tensor._make_wrapper_subclass(cls, shape, **kwargs)  # type: ignore[attr-defined]
+
+    def __init__(
+        self,
+        packed_weight: torch.Tensor,
+        scale: torch.Tensor,
+        zero_point: torch.Tensor,
+        gemlite_kwargs: Dict,
+        _layout: Layout,
+    ):
+        self.packed_weight = packed_weight
+        self.scale = scale
+        self.zero_point = zero_point
+        self.gemlite_kwargs = gemlite_kwargs
+        self._layout = _layout
+        torch._dynamo.config.inline_inbuilt_nn_modules = False
+
+    def __tensor_flatten__(self):
+        return ["packed_weight", "scale", "zero_point"], [
+            self._layout,
+            self.gemlite_kwargs,
+        ]
+
+    @classmethod
+    def __tensor_unflatten__(
+        cls, tensor_data_dict, tensor_attributes, outer_size, outer_stride
+    ):
+        packed_weight, scale, zero_point = (
+            tensor_data_dict["packed_weight"],
+            tensor_data_dict["scale"],
+            tensor_data_dict["zero_point"],
+        )
+        _layout, gemlite_kwargs = tensor_attributes
+        return cls(packed_weight, scale, zero_point, gemlite_kwargs, _layout)
+
+    @classmethod
+    def from_plain(
+        cls,
+        int_data: torch.Tensor,
+        scale: torch.Tensor,
+        zero_point: Optional[torch.Tensor],
+        _layout: Layout,
+    ):
+        from gemlite.core import DType, GemLiteLinearTriton, set_autotune
+
+        assert isinstance(
+            _layout, GemlitePackedLayout
+        ), f"GemliteAQTTensorImpl only works with GemliteLinearTriton but got {_layout}"
+        group_size, bit_width = _layout.group_size, _layout.bit_width
+
+        torch._dynamo.config.inline_inbuilt_nn_modules = False
```

Comment:
nit: do we need to restore the state afterwords

Reply:
i think this is necessary for gemlite to function correctly,